packages/beacon-node/src/sync/utils/downloadByRange.ts

```diff
@@ -528,109 +547,268 @@ export async function validateBlobsByRangeResponse(
+
+        continue;
+      }
+
+      currentSlotSeen.push(columnSidecar);
+
+      if (existingCurrentSlotSeen) {
+        warnings.push(
+          new DownloadByRangeError(
+            {
+              code: DownloadByRangeErrorCode.OUT_OF_ORDER_COLUMNS,
+              slot,
+            },
+            "Existing currentSlotSeen was found, slots delivered out of order"
+          )
+        );
+        // sort as they could be out of order
+        currentSlotSeen.sort((a, b) => a.index - b.index);
+      }
+    } else {
+      // out of order
+      warnings.push(
+        new DownloadByRangeError(
+          {
+            code: DownloadByRangeErrorCode.OUT_OF_ORDER_COLUMNS,
+            slot,
+          },
+          "ColumnSidecars received out of slot order"
+        )
+      );
+
+      const slotColumns = seenColumns.get(slot) ?? [];
+      const lastIndex = slotColumns.at(-1)?.index;
+
+      if (slotColumns.find((c) => c.index === columnSidecar.index)) {
+        warnings.push(
+          new DownloadByRangeError({
+            code: DownloadByRangeErrorCode.DUPLICATE_COLUMN,
+            slot,
+            index: columnSidecar.index,
+          })
+        );
+
+        continue;
+      }
+
+      slotColumns.push(columnSidecar);
+
+      if (lastIndex && lastIndex > columnSidecar.index) {
+        warnings.push(
+          new DownloadByRangeError(
+            {
+              code: DownloadByRangeErrorCode.OUT_OF_ORDER_COLUMNS,
+              slot,
+            },
+            "Column indices out of order within a slot"
+          )
+        );
+        // sort as they could be out of order
+        slotColumns.sort((a, b) => a.index - b.index);
+      }
+
+      seenColumns.set(slot, slotColumns);
+    }
+
+    // make sure last iteration gets set to seenColumns
+    seenColumns.set(currentSlot, currentSlotSeen);
   }
 
-  const warnings: DownloadByRangeError[] = [];
-  // no need to check for columnSidecars.length  vs expectedColumnCount here, will be checked per-block below
-  const requestedColumns = new Set(request.columns);
-  const validateSidecarsPromises: Promise<ValidatedColumnSidecars>[] = [];
-  for (let blockIndex = 0, columnSidecarIndex = 0; blockIndex < dataRequestBlocks.length; blockIndex++) {
-    const {block, blockRoot} = dataRequestBlocks[blockIndex];
+  const validationPromises: Promise<ValidatedColumnSidecars>[] = [];
+
+  for (const {blockRoot, block} of blocks) {
     const slot = block.message.slot;
-    const blockRootHex = toRootHex(blockRoot);
-    // TODO GLOAS: Post-gloas's blobKzgCommitments is not in beacon block body. Need to source it from somewhere else.
-    const blockKzgCommitments = (block as SignedBeaconBlock<ForkPostFulu & ForkPreGloas>).message.body
-      .blobKzgCommitments;
-    const expectedColumns = blockKzgCommitments.length ? request.columns.length : 0;
+    // TODO(gloas): Post-gloas's blobKzgCommitments is not in beacon block body. Need to source it from somewhere else.
```

Comment:
let's keep the todo comments consistent, all other existing gloas todos are like this
```suggestion
    // TODO GLOAS: Post-gloas's blobKzgCommitments is not in beacon block body. Need to source it from somewhere else.
```

Reply:
updated. fwiw I think TODO should be structured like a conventional commit with the "scope" inside parentheses.

Reply:
I don't mind either format, usually I just check what's already commonly used for that specific fork and go with that